Support for the Json error output format of rustc including macro errors

PieterPenninckx opened the pull request:
Add support for the new Json error output format of rustc. Macro errors are supported as well. Fixes #130. The parser for the old format is still there. Based upon the compiler flags (environment variable, actually), the plugin determines which parser to use. In order to use the new parser, the user has to set the environment variable `RUSTFLAGS` to `--error-format json` in the Build Command.

This pull request contains a copy of the minimal-json library, which is copyrighted by EclipseSource and distributed under the MIT license, see plugin_tooling/lib/minimal-json-README.txt for more details.

Commit:
Added dependency on minimal-json and README with details about this.

Commit:
Added a method to check if the Rust compiler is configured to output its errors in the json format.

Commit:
Added tests for the Json output, refactored tests, fixed magic constant, updated copyright statement.

Commit:
Moved regexp-based line parsers for rustc and cargo error output from RustBuildOutputParser to new classes.

Commit:
Add parser for Json error output format.
Add parser for Json error output format of rustc,
use this parser and fix small bug in CargoBuildOutputLineParser.

Commit:
Merge pull request #1 from RustDT/master

Update from RustDT/RustDT

Commit:
Merge https://github.com/PieterPenninckx/RustDT into new-error-output-format

Commit:
Call super constructor from CompositeToolMessageData constructor + small changes
Use super constructor in CompositeToolMessageData
(inner class in RustBuildOutputParser), remove outdated comment and add
copyright notice.

Commit:
Merge https://github.com/RustDT/RustDT into new-error-output-format

Commit:
Add test for parsing macro-related errors in Json output.

Commit:
Show errors in macro expansion at expansion site.

Commit:
Remove unused data from tests

Commit:
Show errors in macros at declaration site as well.

Commit:
Make macro errors more verbose at macro declaration site.

bruno-medeiros commented:
Cool, I shoud have time to review this properly later in the week.  

For starters, regarding minimal-json: interesting choice, I wasn't aware of that library. In https://github.com/GoClipse/goclipse/tree/master/plugin_tooling I used http://www.json.org/ (which has no dependencies as well) But minimal-json claims to be faster, so that's good. Is there any reason you prefered this library? (API, etc, - or was it mainly the performance aspect?)

PieterPenninckx commented:
Criteria for the Json parsing library were:
- license: I wanted something MIT-licensed (or similar) for compatibility with the Eclipse Public License
- API: I wanted something simple where you can "walk the tree" instead of "serialisation/deserialisation" if you get what I mean. Serialisation/deserialisation is handy when your data structure matches the one used in the Json format, but that is not the case.
- something that would fit your criteria regarding size, scope, dependencies, ... I tried to reverse-engineer your criteria from your choice for the Toml parser. I also liked the fact that minimal-json doesn't have dependencies, but that's probably just my personal taste.

Performance was not a criterion for me because I hope that most projects have only a limited amount of errors, so the Json should be quite small.

Then I just picked something; I didn't want to loose too much time comparing libraries.

bruno-medeiros commented:
I still haven't had time to review properly, hopefully next week. 

In the meanwhile, I did some work to be able to integrate minimal-json and other Maven bundles as a proper OSGI dependency, instead of including the jar directly. This will be important for some other RustDT changes (LSP support) that will add dependencies to several 3rd party libraries, not part of Eclipse Platform.

PieterPenninckx commented on plugin_tooling/src-tests/com/github/rustdt/tooling/RustBuildOutputParserClassicTest.java:
```diff
@@ -0,0 +1,169 @@
+import melnorme.utilbox.core.CommonException;
+
+
+public class RustBuildOutputParserClassicTest extends RustBuildOutputParserTest {
```
RustBuildOutputParserTest.java contained all the tests for the output parser. I moved the code specific to the 'classical' format (rustc 1.11 or below) to this file and kept the general code in RustBuildOutputParserTest.java.

PieterPenninckx commented on plugin_tooling/src-tests/com/github/rustdt/tooling/RustBuildOutputParserTest.java:
```diff
@@ -49,135 +29,4 @@ protected static ToolSourceMessage msg(Path path, int line, int column, int endL
 		msg.toString();
 		return msg;
 	}
```
I moved the code that is specific to the "classical" parser (rustc 1.11 or below) to the file RustBuildOutputParserClassicTest.java.

PieterPenninckx commented on plugin_tooling/src/com/github/rustdt/tooling/AbstractRustBuildOutputLineParser.java:
```diff
@@ -0,0 +1,88 @@
+/*******************************************************************************
```
RustBuildOutputParser.java contained code for parsing error output from cargo. Since the error output from cargo is left unchanged in rustc 1.12, I wanted to re-use this code. Since the code for parsing cargo output was a special case of regexp-based parsing, I moved all the regexp-based code out of RustBuildOutputParser.java. This file contains the general, abstract code, CargoRustBuildOutputLineParser.java contains the cargo-specific code (that I also wanted to use in the new parser) and ClassicRustcRustBuildOutput**Line**Parser.java contains the regexp-based code that is still used in the "old" parser.

PieterPenninckx commented on plugin_tooling/src/com/github/rustdt/tooling/RustBuildOutputParser.java:
```diff
@@ -1,21 +1,20 @@
 /*******************************************************************************
```
Differences in this file are because I separated the regexp-based code out of this file, so that I could re-use the regexp-based code for the cargo output.

PieterPenninckx commented:
Ok. I added some notes that explain what changes I made. I hope that helps with reviewing.

bruno-medeiros commented on plugin_tooling/src-tests/com/github/rustdt/tooling/RustBuildOutputParserJsonTest.java:
```diff
@@ -0,0 +1,320 @@
+			listFrom(msg(path("src/main.rs"), 1, 2, 3, 17, WARNING, "BLAH BLAH BLAH")));
+		
+		testParseMessages(buildParser,
+				"{\"message\":\"BLAH BLAH BLAH\",\"level\":\"warning\"," + 
```
This is way too ugly, all that inline Json that then has to escape quotes (since Java doesn't have raw string literals). This should be done in one of two ways:
- Either put the json code in a separate json file, so you don't have to escape quotes, and load it using Java resources (see for example how `getClassResource` is used in `RustAttributeRulesTest`.
- Or, if you wanna keep the json inline with the Java code, use a char other than `"`  for quotes, and then replace it before running the test. See for example `RustRawStringLiteralRule_Test`  which does `source = source.replace('|', '"');`

PieterPenninckx replied:
This has been changed with commit `f75e9ba`. I opted for the separate json files.

bruno-medeiros commented on plugin_tooling/src/com/github/rustdt/tooling/RustBuildOutputParserJson.java:
```diff
@@ -0,0 +1,254 @@
+	// from the method `parseMessageData`.
+	// So we create a buffer of remaining data.
+	protected String lineParsing = null; // The line we are currently working with. 
+	protected Queue<ToolMessageData> buffer; // Invariant: if buffer.isEmpty() then lineParsing == null
```
There is no need to do this. You can overrider doParseToolMessage directly, and detected if it's a json object there, and if it is, read all messages and add all of them straight away using `addBuildMessage`, with not need to keep a list.

PieterPenninckx replied:
This has been changed as you requested with commit `4d1b43a`. It's indeed more simple now.

bruno-medeiros commented on plugin_tooling/src/com/github/rustdt/tooling/RustBuildOutputParserJson.java:
```diff
@@ -0,0 +1,254 @@
+			} else {
+				this.lineParsing = LexingUtils.stringUntilNewline(output);
+				
+				if (this.lineParsing.startsWith("{")) {
```
This seems to work for now but I worry that in the future the JSON output for a single object might span more than one line. Then this code breaks.
Instead, we should create a java.io.Reader subclass that reads from StringCharSource output, and pass that to `Json.parse()`. That willl be much more robust. This Reader subclass should be fairly easy to implement too. (you can put it inside StringCharSource , create a `StringCharSource.toReader()`

PieterPenninckx replied:
The catch for me is that while every output line of rustc is valid JSON, rustc's output as a whole is not. Simplified example to illustrate:

``` json
{"message":"Cannot borrow immutable variable as mutable"}
{"message":"Cannot borrow variable as mutable more than once"}
```

This is not valid JSON, but rather two (since there are two lines) pieces of valid JSON joined together and separated by a newline. The newline character is used to separate two messages. 

Unfortunately, minimal-json cannot handle multiple JSON chunks separated by newlines. Let me illustrate this with an example:

``` java
import com.eclipsesource.json.Json;
import com.eclipsesource.json.JsonValue;
import com.eclipsesource.json.ParseException;
import java.io.StringReader;
import java.io.IOException;

public class Main {
    public static void main(String[] args) {
        StringReader reader = new StringReader("{\"a\":1}\n{\"a\":2}");
        try {
            JsonValue a = Json.parse(reader);
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }
}
```

In this example, I create a Reader subclass that feeds something that is not quite JSON (but in the same style as rustc's output) to minimal-json. Instead of "eating" as much of the input as is valid JSON, `Json.parse()` throws a `com.eclipsesource.json.ParseException: Unexpected character at 2:1`.

In summary:
- `rustc`'s output as a whole is no valid JSON, but pieces of valid JSON separated by newlines
- minimal-json cannot handle this, so
- we split `rustc`'s output into lines (which are right now valid JSON) that we can feed to `Json.parse()`.

If future's `rustc`'s output were to split messages over multiple lines, the current approach would stop working as you noticed. I see different options to solve this:
1. Feed `rustc`'s output as a whole to `Json.parse()`. I believe this is what you proposed (correct me if I'm wrong), but unfortunately `Json.parse()` would throw an exception as illustrated with the example above.
2. Look for an alternative JSON parsing library that only eats as much as is valid JSON, parses that, and allows us to handle the rest (by feeding that to the same parser separately). 
3. Create a pull request for minimal-json to allow eating only as much as is valid JSON.
4. Override minimal-json's `Json.parse()` function (or `com.eclipsesource.json.JsonParser`'s method `parse(Reader reader, int buffersize)`).
5. Somehow detect which newlines are separating JSON parts and which are just whitespace. This can either be done by counting curly brackets (ignoring curly brackets which occur between quotes), or by trying to parse larger and larger pieces of the input until `Json.parse()` throws no exceptions anymore. I'm certainly no fan of the latter.
6. Cross our fingers and hope that `rustc`s output will not split JSON over multiple lines.

I would go for option 6., 3. or 4. What is your preference?

The only other IDE/editor that I found which uses the Json output is VisualRust. If I understood it correctly, they use `Newtonsoft.Json` (C#) which can [only eat as much as is valid json](http://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonReader_SupportMultipleContent.htm).

bruno-medeiros replied:
Oh, minimal-json doesn't support parsing part of a string only? 😲 I didnt' expect that, that's a deal-breaker. I say we go with option 2, namely, let's just use google GSON. It does support parsing just part of the input String.

GSON is a clear choice because, well the performance is not too far from minimal-json, but actually, the performance advantage of mininal-json in this scenario is essentially irrelevant because:
a) the volume of messages from Cargo/Rustc is generally be really low, such that performance will not be a bottleneck.
b) even if we want to care about performance (which we could care, even just for the sake of training our software craftsmanship skills), the correct choice would be to use a JSON streaming API, not one that has to create intermediate JSON objects like JsonObject, JSonArray, etc. And google GSON does provide a streaming API. (and minimal-json doesn't)

So yeah, GSON is best option in either case, let's use it (you don't have to use the streaming API BTW, it's not something worth optimizing now)

bruno-medeiros replied:
See https://github.com/google/gson/blob/master/gson/src/main/java/com/google/gson/JsonParser.java#L56

PieterPenninckx replied:
I have switched to the GSON parsing library now. Still on my TODO list:
- [X] Remove the dependency on the minimal-json library.
- [X] Screen for all kinds of unchecked exceptions in the GSON parsing library. I'm not saying I can find all of them, but it's a little less convenient than the minimal-json library and right now you can get `ClassCastExceptions` and what not if you craft the Json.
- [X] Make sure the continuous integration test succeeds, apparently.

PieterPenninckx replied:
Ok, I have fully switched to the GSON parsing library, and the associated TODO list is cleared.

It seems like the Travis check has failed because the Travis server could not download some Eclipse stuff (timeout). Is there a way to ask Travis to retry? Or do I have to create an empty dummy commit for that?

PieterPenninckx replied:
Looks like only the owner can restart a CI build, so I retriggered with an empty commit.

Commit:
Move inline Json in tests from to separate json files.

Commit:
Add test for the Json parser for ignoring some cargo output.

Commit:
Simplify RustBuildOutputParserJson.
Override doParseToolMessage() which eliminates the need for keeping a queue.

Commit:
Add dependency on the gson library.
For license details, see plugin_tooling/lib/gson-README.txt.
The minimal-json library will be removed when the migration is completed.

Commit:
Switch from minimal-json to gson as Json parsing library.

Commit:
Merge https://github.com/RustDT/RustDT into new-error-output-format

Commit:
Fix compilation problem.

Commit:
Remove dependency on minimal-json.

Commit:
Avoid getting exceptions from GSON parsing + code cleanup.

Commit:
Trigger CI rebuild.

bruno-medeiros commented:
Alright, I'm revewing this, and merging it in.

bruno-medeiros commented:
I've merged this in with a squashed commit and a few minor changes, and some notes for a few other minor follow-up changes (to come in a subsequent commit) 

Thanks for the contribution @PieterPenninckx ! 👍
